File: intrinsic/resources/client/resource_registry_client.py
# Copyright 2023 Intrinsic Innovation LLC

"""Lightweight Python wrapper around the resource registry."""

from __future__ import annotations

from typing import Optional

import grpc
from intrinsic.resources.proto import resource_handle_pb2
from intrinsic.resources.proto import resource_registry_pb2
from intrinsic.resources.proto import resource_registry_pb2_grpc
from intrinsic.util.grpc import error_handling

_RESOURCE_REGISTRY_MAX_PAGE_SIZE = 200


class ResourceRegistryClient:
  """Client library for the resource registry gRPC service."""

  _stub: resource_registry_pb2_grpc.ResourceRegistryStub

  def __init__(self, stub: resource_registry_pb2_grpc.ResourceRegistryStub):
    """Constructs a new ResourceRegistryClient object.

    Args:
      stub: The gRPC stub to be used for communication with the resource
        registry service.
    """
    self._stub = stub

  @classmethod
  def connect(cls, grpc_channel: grpc.Channel) -> ResourceRegistryClient:
    """Connects to a running resource registry.

    Args:
      grpc_channel: Channel to the resource registry gRPC service.

    Returns:
      A newly created instance of the ResourceRegistryClient class.
    """
    stub = resource_registry_pb2_grpc.ResourceRegistryStub(grpc_channel)
    return cls(stub)

  @error_handling.retry_on_grpc_unavailable
  def get_resource_instance(
      self, name: str
  ) -> resource_registry_pb2.ResourceInstance:
    """Returns the resource instance with the given name.

    Args:
      name: Name of the resource instance to return.

    Returns:
      The requested resource instance.
    """
    return self._stub.GetResourceInstance(
        resource_registry_pb2.GetResourceInstanceRequest(name=name)
    )

  def list_all_resource_instances(
      self, *, resource_family_id: Optional[str] = None
  ) -> list[resource_registry_pb2.ResourceInstance]:
    """Retrieves all available resource instances.

    Use with caution: This method retrieves all pages of relevant resource
    instances from the backend and blocks until all results have been retrieved.

    Args:
      resource_family_id: Query and return only resource instances with this
        family id.

    Returns:
      List of resource handles.
    """
    base_request = resource_registry_pb2.ListResourceInstanceRequest()
    if resource_family_id is not None:
      base_request.strict_filter.resource_family_id = resource_family_id

    return self._list_all_resource_instances(base_request)

  def list_all_resource_handles(
      self, *, capability_names: Optional[list[str]] = None
  ) -> list[resource_handle_pb2.ResourceHandle]:
    """Retrieves all handles of all resource instances which have handles.

    Use with caution: This method retrieves all pages of relevant resource
    instances from the backend and blocks until all results have been retrieved.

    Args:
      capability_names: Required capability names. Only handles having (at
        least) all of these capabilities will be returned.

    Returns:
      List of resource handles.
    """
    base_request = resource_registry_pb2.ListResourceInstanceRequest()
    if capability_names is not None:
      base_request.strict_filter.capability_names.extend(capability_names)

    all_instances = self._list_all_resource_instances(base_request)

    return [
        instance.resource_handle
        for instance in all_instances
        if instance.HasField('resource_handle')
    ]

  def batch_list_all_resource_handles(
      self, *, capability_names_batch: list[list[str]]
  ) -> list[list[resource_handle_pb2.ResourceHandle]]:
    """Retrieves all handles for each of the given capability requirements.

    Use with caution: This method retrieves all pages of relevant resource
    instances from the backend and blocks until all results have been retrieved.

    Args:
      capability_names_batch: Required capability names. For each of the given
        capability lists, the result will contain one list of handles which have
        (at least) all of these capabilities.

    Returns:
      One list of resource handles for every given list of capability names.
    """
    if not capability_names_batch:
      return []

    # The resource registry only supports filtering for one set of capabilities
    # at a time. To avoid sending many requests, we get all handles (unfiltered)
    # and then do post-filtering below.
    all_handles = self.list_all_resource_handles()

    result: list[list[resource_handle_pb2.ResourceHandle]] = []
    for required_capability_names in capability_names_batch:
      matching_handles = [
          handle
          for handle in all_handles
          if set(required_capability_names).issubset(handle.resource_data)
      ]
      result.append(matching_handles)

    return result

  def _list_all_resource_instances(
      self,
      base_request: resource_registry_pb2.ListResourceInstanceRequest = (
          resource_registry_pb2.ListResourceInstanceRequest()
      ),
  ) -> list[resource_registry_pb2.ResourceInstance]:
    """Retrieves all resource instances by querying all response pages."""
    request = resource_registry_pb2.ListResourceInstanceRequest()
    request.CopyFrom(base_request)
    request.page_size = _RESOURCE_REGISTRY_MAX_PAGE_SIZE

    result: list[resource_registry_pb2.ResourceInstance] = []
    while True:
      response = self._call_list_resource_instances(request)
      result.extend(response.instances)

      if response.next_page_token:
        request.page_token = response.next_page_token
      else:
        return result

  @error_handling.retry_on_grpc_unavailable
  def _call_list_resource_instances(
      self, request: resource_registry_pb2.ListResourceInstanceRequest
  ) -> list[resource_registry_pb2.ResourceInstance]:
    return self._stub.ListResourceInstances(request)

File: intrinsic/resources/client/resource_registry_client.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/resources/client/resource_registry_client.h"

#include <iterator>
#include <memory>
#include <string>
#include <vector>

#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "absl/time/clock.h"
#include "absl/time/time.h"
#include "grpcpp/client_context.h"
#include "intrinsic/resources/proto/resource_registry.grpc.pb.h"
#include "intrinsic/resources/proto/resource_registry.pb.h"
#include "intrinsic/util/grpc/grpc.h"
#include "intrinsic/util/status/status_conversion_grpc.h"
#include "intrinsic/util/status/status_macros.h"

namespace intrinsic {
namespace resources {

absl::StatusOr<std::unique_ptr<ResourceRegistryClient>>
CreateResourceRegistryClient(absl::string_view grpc_address,
                             absl::Duration timeout,
                             absl::Duration connection_timeout) {
  INTR_ASSIGN_OR_RETURN(
      std::shared_ptr<grpc::Channel> channel,
      CreateClientChannel(grpc_address, absl::Now() + connection_timeout));
  return std::make_unique<ResourceRegistryClient>(
      intrinsic_proto::resources::ResourceRegistry::NewStub(channel), timeout);
}

absl::StatusOr<std::vector<intrinsic_proto::resources::ResourceInstance>>
ResourceRegistryClient::ListResources(
    const intrinsic_proto::resources::ListResourceInstanceRequest::StrictFilter
        &filter) const {
  std::vector<intrinsic_proto::resources::ResourceInstance> resource_instances;
  std::string page_token;
  auto deadline = absl::ToChronoTime(absl::Now() + timeout_);
  do {
    ::grpc::ClientContext context;
    context.set_deadline(deadline);
    intrinsic_proto::resources::ListResourceInstanceRequest req;
    intrinsic_proto::resources::ListResourceInstanceResponse resp;
    req.set_page_token(page_token);
    *req.mutable_strict_filter() = filter;
    INTR_RETURN_IF_ERROR(
        ToAbslStatus(stub_->ListResourceInstances(&context, req, &resp)));
    resource_instances.insert(resource_instances.end(),
                              std::make_move_iterator(resp.instances().begin()),
                              std::make_move_iterator(resp.instances().end()));
    page_token = resp.next_page_token();
  } while (!page_token.empty());

  return resource_instances;
}

absl::StatusOr<intrinsic_proto::resources::ResourceInstance>
ResourceRegistryClient::GetResource(absl::string_view name) const {
  ::grpc::ClientContext context;
  context.set_deadline(absl::ToChronoTime(absl::Now() + timeout_));

  intrinsic_proto::resources::GetResourceInstanceRequest req;
  intrinsic_proto::resources::ResourceInstance instance;
  req.set_name(name);
  INTR_RETURN_IF_ERROR(
      ToAbslStatus(stub_->GetResourceInstance(&context, req, &instance)));
  return instance;
}

}  // namespace resources
}  // namespace intrinsic
